Polish the failure summary

jcazevedo opened the pull request:
This PR attempts to polish the failure summary, i.e. the message that is printed by the exception the user may decide to throw when attempting to derive an object from a `ConfigValue`. It does so by grouping and sorting errors by the keys they originated from. 

Incidentally, this fixes an issue with the keys for errors of Map derivations not being properly concatenated (the derivation was missing a call to `improveFailures`), makes the output of the `WrongType` failure more uniform, by always referring to the appropriate `ConfigValueType`s, and removes the `WrongTypeException` class, which was not being used, and was most likely forgotten.

This fixes #154.

jcazevedo committed:
Polish failure summary

This polishes the failure summary by grouping and sorting errors by the keys
they originated from. Incidentally, this fixes an issue with the keys for errors
of Map derivations not being properly concatenated, and makes the output of the
WrongType failure more uniform, by always referring to the appropriate
ConfigValueTypes.

jcazevedo committed:
Improve message of CollidingKeys

melrief commented on core/src/main/scala/pureconfig/BasicConverters.scala:
```diff
@@ -90,15 +90,15 @@ trait TypesafeConfigConverters extends ConvertHelpers {
     override def from(config: ConfigValue): Either[ConfigReaderFailures, Config] = config match {
       case co: ConfigObject => Right(co.toConfig)
-      case other => fail(WrongType(other.valueType().toString, "ConfigObject", ConfigValueLocation(config), None))
+      case other => fail(WrongType(s"${other.valueType}", s"${ConfigValueType.OBJECT}", ConfigValueLocation(config), None))
```
I'm maybe too picky but when I prefer to use the standard `.toString` instead of string interpolation for one value. Same for all the other occurrences.

jcazevedo replied:
Sure. I've changed it in 9c20d08.

jcazevedo committed:
Remove string interpolation in favor of toString calls

ruippeixotog commented on core/src/main/scala/pureconfig/BasicConverters.scala:
```diff
@@ -90,15 +90,15 @@ trait TypesafeConfigConverters extends ConvertHelpers {
     override def from(config: ConfigValue): Either[ConfigReaderFailures, Config] = config match {
       case co: ConfigObject => Right(co.toConfig)
-      case other => fail(WrongType(other.valueType().toString, "ConfigObject", ConfigValueLocation(config), None))
+      case other => fail(WrongType(other.valueType.toString, ConfigValueType.OBJECT.toString, ConfigValueLocation(config), None))
```
I see you started using `ConfigValueType` instances to refer to expected types in `WrongType`. In order to make `WrongType` more type-safe, why don't you make `expectedType` a `List[ConfigValueType]`? That way you make sure the expected types passed are always valid.

jcazevedo replied:
Makes sense. I modified `WrongType` to only accept `ConfigValueType`s in 6079331.

ruippeixotog commented on core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala:
```diff
@@ -72,6 +72,16 @@ sealed abstract class ConfigReaderFailure {
+  /**
+   * A human-readable description of the failure, along with its location.
+   */
+  def descriptionWithLocation: String = location.fold(description)(_.description + " " + description)
```
I preferred if this was a private function local to `ConfigReaderException`. In my view at least, `ConfigReaderFailure` should be as decoupled from presentation as possible, while all logic for turning "failure" domain models into nice error messages should be in `ConfigReaderException` whenever possible. What do you think?

jcazevedo replied:
I agree. Moved the `descriptionWithLocation` method to the `ConfigReaderException` companion object in 53ede88.

ruippeixotog commented on core/src/main/scala/pureconfig/error/ConfigReaderException.scala:
```diff
@@ -12,42 +12,30 @@ final case class ConfigReaderException[T](failures: ConfigReaderFailures)(implic
     val linesBuffer = mutable.Buffer.empty[String]
     linesBuffer += s"Cannot convert configuration to a value of class ${ct.runtimeClass.getName}. Failures are:"
```
Suggestion: "value of class" is a bit redundant here. It would be nicer if it read "Cannot convert configuration to a ${ct.runtimeClass.getName}", IMO.

jcazevedo replied:
Fixed it in 668d6cb.

ruippeixotog commented on core/src/main/scala/pureconfig/error/ConfigReaderException.scala:
```diff
@@ -12,42 +12,30 @@ final case class ConfigReaderException[T](failures: ConfigReaderFailures)(implic
-      }
+    failuresWithPath.foreach {
+      case (p, failures) =>
+        linesBuffer += s"  with '$p':"
```
Suggestion: "at" seems more natural than "with" here.

jcazevedo replied:
I struggled a bit with words there. Originally I was using "in", but then it didn't make much sense for missing keys, because they're not **in** there (they're missing!). I ended up with "with", but I agree "at" sounds better. Replaced it in 482f56f.

ruippeixotog reviewed (changes requested):
LGTM, and kudos for adding nice specs for this! I requested some changes but none of them are major blockers, feel free to leave the PR as it is if you prefer.

jcazevedo committed:
Merge branch 'master' into polish-failure-summary

jcazevedo committed:
Use ConfigValueType in WrongType

Since we only deal with instances of ConfigValueType in WrongType (converting
them to strings), it makes sense that the expected arguments are
ConfigValueTypes instead of strings.

jcazevedo committed:
Remove redundant wording

jcazevedo committed:
Replace 'with' with 'at'

jcazevedo committed:
Make descriptionWithLocation local to ConfigReaderException

melrief commented:
Thanks @jcazevedo !
